add WG AI Gateway (#8521)
kubernetes/community: wg-ai-gateway/charter.md

  content to detect and block sensitive or malicious prompts.
* **Token Rate Limiting** - enforce rate limiting rules based on token usage to
  control usage and cost.
* **Semantic Routing** - making a routing decision for an inference request
  based on semantic similarity of the request body.
* **Semantic Caching** - Provide caching for inference response based on the
  semantic similarity of prompts.
* **Response Risk** - Define and enforce content safety rules with inference
  response content to detect and block sensitive responses from generative AI
  models.
* **Failure Modes** - How inference routing failures should be handled, what
  failure modes we think are important to cover. For instance this may
  encapsulate fallback and retry policies.
* **Observability** - What standards for metrics and tracing for “AI Gateway”
  features should be standardized, and how?

> **Note**: The above list of features should be considered an example, and
> non-exhaustive. We may not act on all of these, but the purpose is more to
> illustrate the kind of features we will be exploring.

### In Scope

Overall guidance for the WG is to control scope as much as is feasible. The WG

> should avoid AI-specific functionality where it can: instead favoring the

Reviewer: If your primary target is "egress use case" (as described in the background section earlier), why do you need to handle these low-level bits, which iiuc should be handled at serving layer? Why not just focus on general purpose rate limiting, caching, etc? Apologies for asking noob :sweat_smile:  questions, I'm far from being network expert, I'm just trying to better understand.

> should avoid AI-specific functionality where it can: instead favoring the

Reviewer: The current version definitely reads better, thank you!

for load-balancing and routing inference, particular when dealing with inference
_workloads_: When the use case includes local model serving on the cluster, and
routing and load-balancing features _rely on information from the inference
workloads_, this kind of routing falls under the scope of WG Serving.

A good example of this is the [Gateway API Inference Extension (GIE)][gie].
This project came from WG Serving and specifically handles advanced routing and
load-balancing for inference which is informed by metrics and capabilities being
advertised by the model serving platform (e.g. VLLM). In this vein, the GIE is
effectively an alternative to the Kubernetes `Service` API, whereas this WG
means to operate more at the `Gateway` and `HTTPRoute` level.

Use cases which have to interact with the model serving layer for networking
(as described above) are generally out of scope for this WG. If some feature
the WG is working on absolutely must cross this line, the effort MUST be brought
to WG Serving and worked on as a joint effort with them.

[gie]:https://github.com/kubernetes-sigs/gateway-api-inference-extension

## Deliverables

> * A compendium of AI related networking definitions (e.g. "AI Gateway") and a

Reviewer 2: ```suggestion
* A compendium of AI related networking definitions (e.g. "AI Gateway") and
```

could even apply at the Gateway level). For the purposes of this charter,
we’ll refer to features at this level as “AI Gateway” features.

## Scope

The scope of this WG is to define terms like "AI Gateway" in the context of
Kubernetes and propose deliverables that need to be adopted in order to **manage
AI traffic** on Kubernetes, such as:

* **Prompt Guards** - Define and enforce content safety rules for inference
  content to detect and block sensitive or malicious prompts.
* **Token Rate Limiting** - enforce rate limiting rules based on token usage to
  control usage and cost.
* **Semantic Routing** - making a routing decision for an inference request
  based on semantic similarity of the request body.
* **Semantic Caching** - Provide caching for inference response based on the
  semantic similarity of prompts.
* **Response Risk** - Define and enforce content safety rules with inference
  response content to detect and block sensitive responses from generative AI
  models.
* **Failure Modes** - How inference routing failures should be handled, what
  failure modes we think are important to cover. For instance this may
  encapsulate fallback and retry policies.

> * **Observability** - What standards for metrics and tracing for “AI Gateway”

Reviewer 3: What do we hope to do that is not already done by https://opentelemetry.io/docs/specs/semconv/gen-ai/?

> * **Observability** - What standards for metrics and tracing for “AI Gateway”

Reviewer 4: TIL - yeah I think this probably covers it. If we come across anything missing, I imagine we'll make recommendations to that community.

/cc @shaneutt

> * **Observability** - What standards for metrics and tracing for “AI Gateway”

Author: Agreed, I think that probably covers it as well. We wanted to mention telemetry, but we are not trying to standardize network observability in Kubernetes. We want a clean scope and a clean exit, and trying to accomplish this would be a working group all unto itself.

The language used here is not clear about that however, it is misworded, let's fix it up:

```suggestion
* **Observability** - Evaluate and suggest (existing) mechanisms for observability in “AI Gateways”
```

> * **Observability** - What standards for metrics and tracing for “AI Gateway”

Author: Also made a specific note in the non-goals that we're not making new observability standards, LMKWYT

# WG AI Gateway Charter

This charter adheres to the conventions described in the [Kubernetes Charter
README] and uses the Roles and Organization Management outlined in
[wg-governance].

[wg-governance]:https://github.com/kubernetes/community/blob/master/committee-steering/governance/wg-governance.md
[Kubernetes Charter README]:https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md

## Scope

The AI Gateway Working Group focuses on the intersection of AI and
networking, particularly in the context of extending load-balancer, gateway

> and proxy technologies to manage and route traffic for AI Inference.

Reviewer 5: I’m wondering if this WG has some overlap with WG Serving. 

https://github.com/kubernetes/community/blob/bfee8f77de8c0e31c31cf7f90792148bd7be1f07/wg-serving/charter.md?plain=1#L37-L38 

Could you help clarify the scope differences?

Author: In this case it's squarely focused on the networking aspects of inference, not the compute or lifecycle management aspects as in the case of wg-serving. The key point is this one:

```
manage and route traffic for AI Inference.
```

Reviewer: Right, but why the existing WG Serving isn't the right place to cover those topics? The biggest issue as I'm seeing is that you will affect WG Serving work, only if focusing on the networking side of it. So either explicitly listing out the reasons for that extraction or how are you going to collaborate with WG Serving in that scope will be necessary to ensure the work doesn't diverge.

Author: > So either explicitly listing out the reasons for that extraction or how are you going to collaborate with WG Serving in that scope will be necessary to ensure the work doesn't diverge.

I've added WG Serving as an explicit collaborator which we will need to keep looped in for review on proposals.

/cc @ArangoGutierrez @SergeyKanzhelev @terrytangyuan

addition of provisions that help with AI use-cases, but are otherwise normal
networking facilities. Under that guidance, the following is in-scope:

* Providing definitions for networking related AI terms in a Kubernetes
  context.

* Defining important AI networking use-cases for Kubernetes users.

* Determining which common features and capabilities in the "AI Gateway" space
  need to be covered by Kubernetes standards and APIs according to user and
  implementation needs.

* Creating proposals for "AI Gateway" features and capabilities to the
  appropriate sub-projects.

* Propose new sub-projects if existing sub-projects are not sufficient.

### Out of Scope

* Developing whole "AI Gateway" solutions. This group will focus on
  enabling existing and new solutions to be more easily deployed and managed on
  Kubernetes, not adding any new production solutions maintained thereafter by
  upstream Kubernetes.

* Any specific kind of hardware support is generally out of scope.

* This group will not cover the entire spectrum of networking for AI. For
  instance: RDMA networks are generally out of scope.

## Deliverables

> * A compendium of AI related networking definitions (e.g. "AI Gateway") and a

Reviewer 6: is this some sort of stored artifact?

  key use-cases for Kubernetes users.

* Provide a space for collaboration and experimentation to determine the most

>   viable features and capabitilies that Kubernetes should support. If there is

Reviewer 7: small typo
```suggestion
  viable features and capabilities that Kubernetes should support. If there is
```

  strong consensus on any particular ideas, the WG will facilitate and
  coordinate the delivery of proposals in the appropriate areas.

## Stakeholders

* SIG Network

## Roles and Organization Management

This working group adheres to the Roles and Organization Management outlined in
[wg-governance] and opts-in to updates and modifications to [wg-governance].

[wg-governance]:https://github.com/kubernetes/community/blob/master/committee-steering/governance/wg-governance.md

## Exit Criteria

> The WG is done when its deliverables are complete, according to the defined

Reviewer: Have you considered a SIG-Network sub-project?
